crates/router/src/db/authentication.rs

```diff
@@ -0,0 +1,155 @@
+
+#[async_trait::async_trait]
+pub trait AuthenticationInterface {
+    async fn insert_authentication(
+        &self,
+        authentication: storage::AuthenticationNew,
+    ) -> CustomResult<storage::Authentication, errors::StorageError>;
+
+    async fn find_authentication_by_merchant_id_authentication_id(
+        &self,
+        merchant_id: String,
+        authentication_id: String,
+    ) -> CustomResult<storage::Authentication, errors::StorageError>;
+
+    async fn update_authentication_by_merchant_id_authentication_id(
+        &self,
+        previous_state: storage::Authentication,
+        authentication_update: storage::AuthenticationUpdate,
+    ) -> CustomResult<storage::Authentication, errors::StorageError>;
+}
+
+#[async_trait::async_trait]
+impl AuthenticationInterface for Store {
+    async fn insert_authentication(
+        &self,
+        authentication: storage::AuthenticationNew,
+    ) -> CustomResult<storage::Authentication, errors::StorageError> {
+        let conn = connection::pg_connection_write(self).await?;
+        authentication
+            .insert(&conn)
+            .await
+            .map_err(Into::into)
+            .into_report()
+    }
+
+    async fn find_authentication_by_merchant_id_authentication_id(
+        &self,
+        merchant_id: String,
+        authentication_id: String,
+    ) -> CustomResult<storage::Authentication, errors::StorageError> {
+        let conn = connection::pg_connection_write(self).await?;
+        storage::Authentication::find_by_merchant_id_authentication_id(
+            &conn,
+            &merchant_id,
+            &authentication_id,
+        )
+        .await
+        .map_err(Into::into)
+        .into_report()
+    }
+
+    async fn update_authentication_by_merchant_id_authentication_id(
+        &self,
+        previous_state: storage::Authentication,
+        authentication_update: storage::AuthenticationUpdate,
+    ) -> CustomResult<storage::Authentication, errors::StorageError> {
+        let conn = connection::pg_connection_write(self).await?;
+        storage::Authentication::update_by_merchant_id_authentication_id(
+            &conn,
+            previous_state.merchant_id,
+            previous_state.authentication_id,
+            authentication_update,
+        )
+        .await
+        .map_err(Into::into)
+        .into_report()
+    }
+}
+
+#[async_trait::async_trait]
+impl AuthenticationInterface for MockDb {
+    async fn insert_authentication(
+        &self,
+        authentication: storage::AuthenticationNew,
+    ) -> CustomResult<storage::Authentication, errors::StorageError> {
+        let mut authentications = self.authentications.lock().await;
+        if authentications.iter().any(|authentication_inner| {
+            authentication_inner.authentication_id == authentication.authentication_id
+        }) {
+            Err(errors::StorageError::DuplicateValue {
+                entity: "authentication_id",
+                key: None,
```

Comment:
```suggestion
                entity: "authentication",
                key: Some(authentication.authentication_id),
```

I believe these would be more suitable.